Sources/PackageLoading/ManifestLoader.swift

```diff
@@ -579,8 +579,27 @@ public final class ManifestLoader: ManifestLoaderProtocol {
                     return
                 }
 
-                // Pass the fd in arguments.
-                cmd = [file.pathString, "-fileno", "1"]
+                // Open a pipe to which the JSON representation of the manifest should be written.
+                let jsonOutputPipe = Pipe()
+
+                // Set up to asynchronously collect data from the pipe.
+                var jsonOutputData = Data()
+                let jsonOutputGroup = DispatchGroup()
```

Comment:
Is it possible to avoid this if we use readDataToEndOfFile method after closing the write end?

Reply:
No, that is not possible.  That can cause hangs if the amount of data overflows the pipe buffer.  I've been bitten by that when using a similar pattering (with the same APIs) a different codebase.

My original version for this wrote to a temporary file but it seems almost as straightforward to use the pipe, since we would need to open the file for writing anyway in order to pass the file descriptor.

Reply:
Is the concern that the code is complicated due to the use of dispatch groups, or is there a reliability or performance concern?  There is precedent (in Workspace, I believe) for using a dispatch group (though a dispatch semaphore, which would be the other option, is not used anywhere else in SwiftPM, as far as I could tell).

Reply:
I'm wondering if its worth hoisting that pattern into a function to make stand out, document it, and allowing us to reuse it. It took me a few reads to understand what's going on and I think it would make more sense on its own. Something like:

```
extension Pipe {
    func read(from write: (FileHandle) -> Void) -> Data {
        var data = Data()
        let group = DispatchGroup()
        group.enter()
        fileHandleForReading.readabilityHandler = { (fileHandle: FileHandle) -> Void in
            let newData = fileHandle.availableData
            if newData.isEmpty {
                fileHandleForReading.readabilityHandler = nil
                group.leave()
            } else {
                data.append(newData)
            }
        }

        write(fileHandleForWriting)

        fileHandleForWriting.closeFile()
        group.wait()
        return data
    }
}
```

Reply:
It's a good idea, but it would also be a little bit inverted for the main call from the client side to be to `Pipe.read(from:)` when the main thing that's happening, semantically, is to invoke a subprocess.

What we should ideally do is to extend `popen()` to support more output pipes than just stderr and stdout.  A more flexible implementation would use a single dispatch group inside the handler that waits for output from all the pipes (stdout and stderr included), where it also waits for the completion of the subprocess.

But reworking `popen()` seemed like to big of a change to fix the immediate issue.  It's something we should schedule for the future.

Perhaps we should go back to writing to a file, to make the code simpler — even if it's marginally slower, it's probably in the noise.

Reply:
I am mostly concerned about the extra complexity of this code. One potential option is passing a temporary file path that the PackageDescription library can write into and then we can read it back here. We're already inside a temporary directory closure so it should be straightforward to construct and pass a temporary path using the `tmpDir` variable in this scope.

Reply:
My first implementation of this did essentially what the original interpreter code did, which was to open a new file that temporary directory and then pass a file descriptor in `-fileno`.  I'd prefer to leave the door open for using a pipe in the future, but can change the argument to a file instead.  Then if we end up using a pipe later, we can always change it back since it's a private implementation detail.

I'll revise the PR.

Reply:
I've updated the PR to use a simple file.